velox/expression/Expr.cpp

```diff
@@ -852,17 +852,17 @@ void Expr::evaluateSharedSubexpr(
     VectorPtr& result,
     TEval eval) {
   // Captures the inputs referenced by distinctFields_.
-  std::vector<const BaseVector*> expressionInputFields;
+  InputForSharedResults expressionInputFields;
   for (auto* field : distinctFields_) {
-    expressionInputFields.push_back(
-        context.getField(field->index(context)).get());
+    expressionInputFields.addInput(context.getField(field->index(context)));
   }
 
   // Find the cached results for the same inputs, or create an entry if one
   // doesn't exist.
   auto sharedSubexprResultsIter =
       sharedSubexprResults_.find(expressionInputFields);
-  if (sharedSubexprResultsIter == sharedSubexprResults_.end()) {
+  if (sharedSubexprResultsIter == sharedSubexprResults_.end() ||
+      sharedSubexprResultsIter->first.isExpired()) {
```

Comment:
if the key expired then L873 will fail (since insert only works if the key does not exist) and we might end up reusing the result. Therefore, we either need to remove the key in that case OR overwrite its value with an empty sharedResult()

Reply:
Thanks @bikramSingh91 for pointing out the issue. Updated the diff accordingly.